common/app/conf/Filters.scala

```diff
@@ -67,6 +67,20 @@ object SurrogateKeyFilter extends Filter with ExecutionContexts {
   }
 }
 
+object AmpFilter extends Filter with ExecutionContexts with implicits.Requests {
+  override def apply(nextFilter: (RequestHeader) => Future[Result])(request: RequestHeader): Future[Result] = {
+    if (request.isAmp) {
+      val domain = request.headers.get("Origin").fold("https://" + request.domain)(_)
```

Comment:
I like fold, but you are right that it isn't the most intuitive. Happy to change!